internal/command/clusters/uninstall.go

```diff
@@ -0,0 +1,218 @@
+	"github.com/jetstack/jsctl/internal/kubernetes"
+	"github.com/jetstack/jsctl/internal/kubernetes/clients"
+)
+
+func Uninstall(run types.RunFunc, kubeConfigPath *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "uninstall",
+		Short: "Contains commands to check a cluster before the uninstallation of Jetstack Secure software",
+	}
+
+	cmd.AddCommand(verify(run, *kubeConfigPath))
+
+	return cmd
+}
+
+func verify(run types.RunFunc, kubeConfigPath string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "verify",
+		Short: "Check that a cluster is ready to have Jetstack Software uninstalled",
+		Args:  cobra.MatchAll(cobra.ExactArgs(0)),
+		Run: run(func(ctx context.Context, args []string) error {
+			kubeCfg, err := kubernetes.NewConfig(kubeConfigPath)
+			if err != nil {
+				return err
+			}
+
+			// first, we need to check that secrets containing certs are safe from
+			// garbage collection
+			messagesOwnedSecrets := []string{}
+			secretsClient, err := clients.NewGenericClient[*corev1.Secret, *corev1.SecretList](
+				&clients.GenericClientOptions{
+					RestConfig: kubeCfg,
+					APIPath:    "/api/",
+					Group:      corev1.GroupName,
+					Version:    corev1.SchemeGroupVersion.Version,
+					Kind:       "secrets",
+				},
+			)
+
+			var secretsList corev1.SecretList
+			err = secretsClient.List(ctx, &clients.GenericRequestOptions{}, &secretsList)
+
+			for i := range secretsList.Items {
+				secret := &secretsList.Items[i]
+
+				hasCertificateOwnerRef := false
+				for _, ownerRef := range secret.OwnerReferences {
+					if ownerRef.Kind == "Certificate" {
+						hasCertificateOwnerRef = true
+						break
+					}
+				}
+
+				if hasCertificateOwnerRef {
+					messagesOwnedSecrets = append(messagesOwnedSecrets, fmt.Sprintf("%s/%s has certificate owner ref", secret.Namespace, secret.Name))
+				}
+			}
+
+			// next, check if any certificates are about to be renewed
+			messagesUpcomingRenewals := []string{}
+			upcomingCerts := [][]string{}
+			certificateClient, err := clients.NewCertificateClient(kubeCfg)
+			if err != nil {
+				return fmt.Errorf("error creating certificate client: %s", err)
+			}
+
+			var certificates certmanagerv1.CertificateList
+			err = certificateClient.List(
+				ctx,
+				&clients.GenericRequestOptions{},
+				&certificates,
+			)
+			if err != nil {
+				return fmt.Errorf("error listing certificates: %s", err)
+			}
+			warnBuffer, err := time.ParseDuration("1h")
+			if err != nil {
+				return fmt.Errorf("error parsing duration, this is a bug: %s", err)
+			}
+			for _, c := range certificates.Items {
+				if c.Status.RenewalTime != nil {
+					if time.Now().Add(warnBuffer).After(c.Status.RenewalTime.Time) {
+						diff := c.Status.RenewalTime.Sub(time.Now())
+						messagesUpcomingRenewals = append(
+							messagesUpcomingRenewals,
+							fmt.Sprintf(
+								"%s/%s will be renewed soon (%s)",
+								c.Namespace,
+								c.Name,
+								diff.String(),
```

Comment:
This will be negative for any certs that are currently already being renewed or perhaps have failed renewal a couple of times. Could we separate those certs out, to make this less confusing?